go-controller/pkg/ovn/controller/apbroute/external_controller_policy.go

```diff
@@ -436,519 +311,154 @@ func (m *externalPolicyManager) calculateAnnotatedPodGatewayIPsForNamespace(targ
+// getPolicyConfigAndUpdatePolicyRefs lists and updates all referenced objects for a given policy and returns
+// routePolicyConfig to perform an update.
+// This function should be the only one that lists referenced objects, and updates policyReferencedObjects atomically.
+func (m *externalPolicyManager) getPolicyConfigAndUpdatePolicyRefs(policy *adminpolicybasedrouteapi.AdminPolicyBasedExternalRoute,
+	updateRefs bool) (*routePolicyConfig, error) {
 	var (
 		errors []error
 	)
 	staticGWInfo, err := m.processStaticHopsGatewayInformation(policy.Spec.NextHops.StaticHops)
 	if err != nil {
-		errors = append(errors, err)
+		errors = append(errors, fmt.Errorf("failed to process statis GW: %w", err))
 	}
-	if len(staticGWInfo) > 0 {
+	if staticGWInfo.Len() > 0 {
 		klog.V(5).Infof("Found static hops for policy %s:%+v", policy.Name, staticGWInfo)
 	}
-	dynamicGWInfo, err := m.processDynamicHopsGatewayInformation(policy.Spec.NextHops.DynamicHops)
-	if err != nil {
-		errors = append(errors, err)
-	}
-	if len(dynamicGWInfo) > 0 {
-		klog.V(5).Infof("Found dynamic hops for policy %s:%+v", policy.Name, dynamicGWInfo)
-	}
-	if len(errors) > 0 {
-		return nil, kerrors.NewAggregate(errors)
-	}
-	return &routePolicy{
-		targetNamespacesSelector: &policy.Spec.From.NamespaceSelector,
-		staticGateways:           staticGWInfo,
-		dynamicGateways:          dynamicGWInfo,
-	}, nil
 
-}
-
-func (m *externalPolicyManager) processExternalRoutePolicies(externalRoutePolicies []*adminpolicybasedrouteapi.AdminPolicyBasedExternalRoute) ([]*routePolicy, error) {
-	routePolicies := make([]*routePolicy, 0)
-	for _, erp := range externalRoutePolicies {
-		processedPolicies, err := m.processExternalRoutePolicy(erp)
-		if err != nil {
-			return nil, err
-		}
-		routePolicies = append(routePolicies, processedPolicies)
-	}
-	return routePolicies, nil
-}
+	// take a lock before listing any objects
+	m.policyReferencedObjectsLock.Lock()
+	defer m.policyReferencedObjectsLock.Unlock()
 
-// calculatePolicyDifferences determines the differences between two policies in terms of namespaces where the policy applies, and the differences in static and dynamic hops.
-// The return values are the namespaces, static hops and dynamic hops that are in the first policy but not in the second instance.
-func (m *externalPolicyManager) calculatePolicyDifferences(policy1, policy2 *adminpolicybasedrouteapi.AdminPolicyBasedExternalRoute) (sets.Set[string], []*adminpolicybasedrouteapi.StaticHop, []*adminpolicybasedrouteapi.DynamicHop, error) {
-	mismatchingNamespaces, err := m.calculateNamespaceSelectorDifferences(&policy1.Spec.From.NamespaceSelector, &policy2.Spec.From.NamespaceSelector)
+	dynamicGWInfo, gwNamespaces, gwPods, err := m.processDynamicHopsGatewayInformation(policy.Spec.NextHops.DynamicHops)
 	if err != nil {
-		return nil, nil, nil, err
+		errors = append(errors, fmt.Errorf("failed to process dynamic GW: %w", err))
```

Comment:
why not return now on error rather than do it after processing the namespace selector?

Reply:
it was in the original `processExternalRoutePolicy` function, and I thought that was to get all spec-related errors, e.g. if we fail to parse multiple fields we can return it at once, and policy spec can be fixed at once instead of fixing one error, then updating, and only getting the next error after that. 
But now when I checked what may fail on that path, I think we only have spec-related errors in staticGW spec, so no reason to aggregate error?

Reply:
It might have been an early incarnation of the code where I tried to aggregate the errors following an example elsewhere in the code. But in hindsight, I would now say that I see no need to continue executing once an error is found since we leverage on the requeuing of the event to reprocess it later on.
In any case, not a big thing if you prefer to aggregate errors to identify as many problems as possible before returning.

Reply:
I actually have updated it already, just wanted to check with you if it had some special purpose :)